logging/impl_test.go

```diff
@@ -420,3 +422,126 @@ func TestLoggingWithFields(t *testing.T) {
+					},
+				},
+			},
+			"these are not the droids you are looking for obi undefined",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actualStringification := tc.logEntry.String()
+			test.That(t, actualStringification, test.ShouldEqual, tc.expectedStringification)
+		})
+	}
+}
+
+func TestLoggingDeduplication(t *testing.T) {
+	// Create a logger object that will write to the `notStdout` buffer.
+	notStdout := &bytes.Buffer{}
+	logger := &impl{
+		name:           "impl",
+		level:          NewAtomicLevelAt(DEBUG),
+		appenders:      []Appender{NewWriterAppender(notStdout)},
+		registry:       newRegistry(),
+		testHelper:     func() {},
+		dedupNoisyLogs: true,
+		// Initialize recent message fields as `NewLogger` would.
+		recentMessageCounts:      make(map[string]int),
+		recentMessageEntries:     make(map[string]LogEntry),
+		recentMessageWindowStart: time.Now(),
+	}
+
+	// Artificially lower noisy message window for testing.
+	originalNoisyMessageWindowDuration := noisyMessageWindowDuration
+	noisyMessageWindowDuration = 500 * time.Millisecond
+	defer func() {
+		noisyMessageWindowDuration = originalNoisyMessageWindowDuration
+	}()
+
+	// Log 4 identical messages (same sublogger, messages, and fields) in quick
+	// succession. Sleep for noisy message window duration. Assert that a final,
+	// separate log is an aggregation log.
+	identicalMsg := "identical message"
+	loggerWith := logger.WithFields("key", "value")
+	for range 3 {
+		loggerWith.Info(identicalMsg)
+		assertLogMatches(t, notStdout,
+			`2023-10-30T13:19:45.806Z	INFO	impl	logging/impl_test.go:132	identical message	{"key":"value"}`)
+	}
+	loggerWith.Info(identicalMsg) // not output due to being noisy
+	time.Sleep(noisyMessageWindowDuration)
+	loggerWith.Info("foo") // log arbitrary message to force output of aggregated message
+	assertLogMatches(t, notStdout,
+		`2023-10-30T13:19:45.806Z	INFO	impl	logging/impl_test.go:132	Message logged 4 times in past 500ms: identical message	{"key":"value"}`)
+	assertLogMatches(t, notStdout,
+		`2023-10-30T13:19:45.806Z	INFO	impl	logging/impl_test.go:132	foo	{"key":"value"}`)
+
+	// Assert aggregation resets after sleep (same aggregation occurs again.)
+	for range 3 {
+		loggerWith.Info(identicalMsg)
+		assertLogMatches(t, notStdout,
+			`2023-10-30T13:19:45.806Z	INFO	impl	logging/impl_test.go:132	identical message	{"key":"value"}`)
+	}
+	loggerWith.Info(identicalMsg) // not output due to being noisy
+	time.Sleep(noisyMessageWindowDuration)
+	loggerWith.Info("foo") // log arbitrary message to force output of aggregated message
+	assertLogMatches(t, notStdout,
+		`2023-10-30T13:19:45.806Z	INFO	impl	logging/impl_test.go:132	Message logged 4 times in past 500ms: identical message	{"key":"value"}`)
+	assertLogMatches(t, notStdout,
+		`2023-10-30T13:19:45.806Z	INFO	impl	logging/impl_test.go:132	foo	{"key":"value"}`)
+
+	// TODO(benji): Add the following assertions to test more deduplication logic.
```

Comment:
Still working on testing a bit, but thought I'd throw this into review before the holiday.